README
Dist::Zilla::Plugin::Beam::Connector

This module aims to allow "Dist::Zilla" to use plugins using
"Beam::Event" and "Beam::Emitter", and perhaps reduce the need for
massive amounts of composition and role application proliferating
"CPAN".

This is in lieu of a decent dependency injection system, and is
presently relying on "Dist::Zilla" to load and construct the plugins
itself, and then you just connect the plugins together informally,
without necessitating each plugin be specifically tailored to the
recipient.

Hopefully, this may also give scope for non-"dzil" plugins being
loadable into memory some day, and allowing message passing of events to
those plugins. ( Hence, the "plugin:" prefix )

A Real World Example of what a future could look like?

  [GatherDir]

  [Test::Compile]

  [Beam::Connector]
  on = plugin:GatherDir#collect => plugin:Test::Compile#generate_test

"GatherDir" in this example would build a mutable tree of files, attach
them to an event "::GatherDir::Tree", and pass that event to
"Test::Compile#generate_test", which would then add ( or remove, or
mutate ) any files in that tree.

Tree state mutation then happens in order of prescription, in the order
given by the various "on" declarations.

Thus, a single plugin can be in 2 places in the same logical stage.

  [Beam::Connector]
  on = plugin:GatherDir#collect => plugin:Test::Compile#generate_test
  ; lots more collectors here
  on = plugin:GatherDir#collect => plugin:Test::Compile#finalize_test

Whereas presently, order of affect is either governed by:

*   phase - where you can add but not remove or mutate, mutate but not
    add or remove, remove, but not add or mutate

*   plugin order - where a single plugin cant be both early in a single
    phase and late

If that example is not convincing enough for you, consider all the
different ways there are presently for implementing "[MakeMaker]". If
you're following the standard logic its fine, but as soon as you set out
of the box, you have a few things you're going to have to do instead:

*   Subclass "MakeMaker" in some way

*   Re-implement "MakeMaker" in some way

*   Fuss a lot with phase ordering and then inject code in the "File"
    that "MakeMaker" generates.

These approaches all work, but they're an open door to everyone
re-implementing the same thing thousands of times over.

  [MakeMaker]

  [DynamicPrereqs]
  -phases = none

  [Beam::Connector]
  on = plugin:MakeMaker#collect_augments => plugin:DynamicPrereqs#inject_augments

"MakeMaker" here can just create an "event", pass it to
"DynamicPrereqs", "DynamicPrereqs" can inject its desired content into
the "event", and then "MakeMaker" can integrate the injected events at
"wherever" the right place for them is.

This is much superior to scraping the generated text file and injecting
events at a given place based on a "RegEx" match.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Dist-Zilla-Plugin-Beam-Connector

0.001004 2017-03-07T07:56:05Z 7dcbe59

0.001003 2017-03-07T07:53:59Z cf919e4
 [Bugfix]
 - Ensure Beam::Wire doesn't fail by passing it a stringifed File instead of an object
 - Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - test: ↓1

0.001002 2016-04-30T20:31:19Z 91ceba7
 - Include a few missing testing dependencies that got hidden in the example

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1
 - test: +5

0.001001 2016-04-29T07:36:16Z 75d0f96
 - Revised and improved recommendations around namespacing and packaging

0.001000 2016-04-29T06:41:35Z db240bf
 - First version.


examples/neo-makemaker/inc/My/NeoMakeMaker.pm
use 5.008;    # utf8
use strict;
use warnings;

package inc::My::NeoMakeMaker;

# ABSTRACT: An example MakeMaker with Beam

# AUTHORITY

use Moose qw( around extends with has );
extends 'Dist::Zilla::Plugin::MakeMaker';
with 'Beam::Emitter';
use List::Util qw( first );

# PS: This example shows exactly how much you need to piss around with EUMM
# just to get some basic stuff done.

has 'static_attribution' => (
    isa => 'Bool',
    is  => 'ro',
);

around write_makefile_args => sub {
    my ( $orig, $self, @args ) = @_;
    my $args  = $self->$orig(@args);
    my $event = $_[1]->emit(
        'augment_makefile_args' => class => 'My::MakeMaker::Args',
        args                    => $args
    );
    return $event->args;
};

my $template = q!# This file was automatically generated by {{ $generated_by }}.
use strict;
use warnings;
 
{{ $perl_prereq ? qq[use $perl_prereq;] : ''; }}
 
use ExtUtils::MakeMaker{{ defined $eumm_version && 0+$eumm_version ? ' ' . $eumm_version : '' }};
{{ $share_dir_code{preamble} || '' }}
my {{ $WriteMakefileArgs }}
 
my {{ $fallback_prereqs }}

XXXPRELUDEXXX
 
unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}
 
delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };
 
WriteMakefile(%WriteMakefileArgs);
{{ $share_dir_code{postamble} || '' }}!;

sub gather_files {
    my ($self) = @_;

    require Dist::Zilla::File::InMemory;

    my $event =
      $self->emit( 'generate_prelude', class => 'My::MakeMaker::Prelude' );

    $self->log_debug( [ "Prelude was <%s>", $event->prelude ] );
    my $templ = $template;
    $templ =~ s{XXXPRELUDEXXX}{ $event->prelude }sex;

    my $file = Dist::Zilla::File::InMemory->new(
        {
            name    => 'Makefile.PL',
            content => $templ,          # template evaluated later
        }
    );

    $self->add_file($file);
    return;
}

sub setup_installer {
    my ($self) = @_;

    my $write_makefile_args = $self->write_makefile_args;

    $self->__write_makefile_args($write_makefile_args);    # save for testing

    my $perl_prereq = $write_makefile_args->{MIN_PERL_VERSION};

    my $dumped_args =
      $self->_dump_as( $write_makefile_args, '*WriteMakefileArgs' );

    my $file = first { $_->name eq 'Makefile.PL' } @{ $self->zilla->files };

    $self->log_debug( ['updating contents of Makefile.PL in memory'] );

    my $attribution =
      $self->static_attribution
      ? ref($self)
      : sprintf( "%s v%s", ref($self), $self->VERSION || '(dev)' );

    my $event = $self->emit(
        'augment_template_args',
        class => 'My::MakeMaker::TemplateArgs',
        args  => {
            eumm_version      => \( $self->eumm_version ),
            perl_prereq       => \$perl_prereq,
            share_dir_code    => $self->share_dir_code,
            fallback_prereqs  => \( $self->fallback_prereq_pm ),
            WriteMakefileArgs => \$dumped_args,
            generated_by      => \$attribution,
        }
    );
    my $content = $self->fill_in_string( $file->content, $event->args, );

    $file->content($content);

    return;
}

__PACKAGE__->meta->make_immutable;
no Moose;

package My::MakeMaker::Args;

use Moose qw( extends has );

extends 'Beam::Event';

has 'args' => ( is => 'ro', required => 1 );

__PACKAGE__->meta->make_immutable;

no Moose;

package My::MakeMaker::Prelude;

use Moose qw( extends has );

extends 'Beam::Event';

has 'prelude' => ( is => 'rw', default => '' );

__PACKAGE__->meta->make_immutable;

no Moose;

package My::MakeMaker::TemplateArgs;

use Moose qw( extends has );

extends 'Beam::Event';

has 'args' => ( is => 'ro', required => 1 );

__PACKAGE__->meta->make_immutable;

no Moose;

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Connect events to listeners in Dist::Zilla plugins.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-Beam-Connector",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Plugin::Beam::Connector",
  "PREREQ_PM" => {
    "Beam::Wire" => 0,
    "Carp" => 0,
    "Dist::Zilla::Role::Plugin" => 0,
    "Moose" => 0,
    "MooseX::LazyRequire" => 0,
    "Path::Tiny" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Beam::Emitter" => 0,
    "Beam::Event" => 0,
    "Dist::Zilla::File::InMemory" => 0,
    "Dist::Zilla::Plugin::MakeMaker" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "List::Util" => 0,
    "Test::DZil" => 0,
    "Test::More" => 0,
    "constant" => 0,
    "lib" => 0
  },
  "VERSION" => "0.001004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/example/*.t"
  }
);


my %FallbackPrereqs = (
  "Beam::Emitter" => 0,
  "Beam::Event" => 0,
  "Beam::Wire" => 0,
  "Carp" => 0,
  "Dist::Zilla::File::InMemory" => 0,
  "Dist::Zilla::Plugin::MakeMaker" => 0,
  "Dist::Zilla::Role::Plugin" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "List::Util" => 0,
  "Moose" => 0,
  "MooseX::LazyRequire" => 0,
  "Path::Tiny" => 0,
  "Test::DZil" => 0,
  "Test::More" => 0,
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
